From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: zhangzujian <zhangzujian.7@gmail.com>
Date: Tue, 22 Sep 2026 07:00:00 +0000
Subject: [PATCH] northd: restore bcast_arp_req_flood with default false

Restore the Kube-OVN NB option that skips L2 flooding of broadcast ARP
and IPv6 ND requests and unicasts localnet ARP for known LSP IPs. Default
the option to false so large underlay broadcast domains do not hit OVS'
4096 resubmit limit. Skip router LSPs so priority-110 forwarding does not
override the priority-105 drop of ARP/ND to LRP addresses from localnet.

Signed-off-by: zhangzujian <zhangzujian.7@gmail.com>
---

diff --git a/northd/en-global-config.c b/northd/en-global-config.c
index bde7806..cd14778 100644
--- a/northd/en-global-config.c
+++ b/northd/en-global-config.c
@@ -710,6 +710,11 @@ check_nb_options_out_of_sync(
         return true;
     }
 
+    if (config_out_of_sync(&nb->options, &config_data->nb_options,
+                           "bcast_arp_req_flood", false)) {
+        return true;
+    }
+
     if (config_out_of_sync(&nb->options, &config_data->nb_options,
                            "install_ls_lb_from_router", false)) {
         return true;
diff --git a/northd/northd.c b/northd/northd.c
index 2485ab5..9e525df 100644
--- a/northd/northd.c
+++ b/northd/northd.c
@@ -119,6 +119,8 @@ static bool ls_ct_skip_dst_lport_ips = false;
 
 #define MAX_OVN_NF_GROUP_IDS 256
 
+static bool bcast_arp_req_flood = false;
+
 /* Due to various hard-coded priorities need to implement ACLs, the
  * northbound database supports a smaller range of ACL priorities than
  * are available to logical flows.  This value is added to an ACL
@@ -10462,6 +10464,12 @@ build_lswitch_lflows_l2_unknown(struct ovn_datapath *od,
                       "outport == \"none\"",
                       "outport = \""MC_UNKNOWN "\"; output;",
                       lflow_ref);
+        if (!bcast_arp_req_flood) {
+            ovn_lflow_add(lflows, od, S_SWITCH_IN_L2_LKUP, 90,
+                          "(eth.bcast && arp.op == 1) || (eth.mcast && nd_ns)",
+                          "next;",
+                          lflow_ref);
+        }
     } else {
         ovn_lflow_add(lflows, od, S_SWITCH_IN_L2_UNKNOWN, 50,
                       "outport == \"none\"", debug_drop_action(), lflow_ref,
@@ -11028,6 +11035,53 @@ build_lswitch_arp_nd_responder_default(struct ovn_datapath *od,
                   lflow_ref);
 }
 
+/* Ingress table 24: forward ARP/ND request from localnet/vtep
+ * to logical switch port for known IPs.
+ * (priority 110)*/
+static void
+build_lswitch_arp_nd_forward_for_unknown_ips(struct ovn_port *op,
+                                             struct lflow_table *lflows,
+                                             struct ds *actions,
+                                             struct ds *match)
+{
+    if (bcast_arp_req_flood) {
+        return;
+    }
+    if (!op->od->nbs) {
+        return;
+    }
+    if (is_vlan_transparent(op->od)) {
+        return;
+    }
+    if (lsp_is_router(op->nbsp)) {
+        return;
+    }
+
+    for (size_t i = 0; i < vector_len(&op->od->localnet_ports); i++) {
+        struct ovn_port *localnet_port = vector_get(
+            &op->od->localnet_ports, i, struct ovn_port *);
+        if (!localnet_port->nbsp) {
+            continue;
+        }
+
+        for (size_t j = 0; j < op->n_lsp_addrs; j++) {
+            for (size_t k = 0; k < op->lsp_addrs[j].n_ipv4_addrs; k++) {
+                ds_clear(match);
+                ds_put_format(match,
+                              "inport == %s && eth.bcast && "
+                              "arp.op == 1 && arp.tpa == %s",
+                              localnet_port->json_key,
+                              op->lsp_addrs[j].ipv4_addrs[k].addr_s);
+                ds_clear(actions);
+                ds_put_format(actions, "outport = %s; output;",
+                              op->json_key);
+                ovn_lflow_add(lflows, op->od, S_SWITCH_IN_ARP_ND_RSP, 110,
+                              ds_cstr(match), ds_cstr(actions), op->lflow_ref);
+            }
+        }
+    }
+}
+
 /* Ingress table 24: ARP/ND responder for service monitor source ip.
  * (priority 110)*/
 static void
@@ -19920,6 +19974,7 @@ build_lswitch_and_lrouter_iterate_by_lsp(struct ovn_port *op,
     build_mirror_lflows(op, ls_ports, lflows);
     build_lswitch_port_sec_op(op, lflows, actions, match);
     build_lswitch_learn_fdb_op(op, lflows, actions, match);
+    build_lswitch_arp_nd_forward_for_unknown_ips(op, lflows, actions, match);
     build_lswitch_from_localnet_op(op, lflows, match);
     build_lswitch_arp_nd_responder_known_ips(op, lflows, ls_ports,
                                              meter_groups, actions, match);
@@ -21407,5 +21462,7 @@ ovnnb_db_run(struct northd_input *input_data,
 
     use_ct_inv_match = smap_get_bool(input_data->nb_options,
                                      "use_ct_inv_match", true);
+    bcast_arp_req_flood = smap_get_bool(input_data->nb_options,
+                                        "bcast_arp_req_flood", false);
     acl_ct_translation = smap_get_bool(input_data->nb_options,
                                        "acl_ct_translation", false);
